### PR TITLE
Big set scenario

### PR DESCRIPTION
**Pull request by aduprat**
The checks of the scenario may be enhanced.

### CODE REVIEW

File: src/main/scala-2.11/org/apache/james/gatling/jmap/JmapMailboxes.scala [hunk 1]
```diff
@@ -5,6 +5,10 @@ import io.gatling.http.Predef._
 import io.gatling.http.check.HttpCheck
 import org.apache.james.gatling.utils.JmapChecks
 import org.apache.james.gatling.utils.RetryAuthentication.execWithRetryAuthentication
+import org.apache.james.gatling.utils.RandomStringGenerator
+
+case class Id(id: String = RandomStringGenerator.randomString)
```
Comment (mbaechler) [inline src/main/scala-2.11/org/apache/james/gatling/jmap/JmapMailboxes.scala]: it should be related to a Mailbox case class IMO.

Comment (mbaechler) [inline src/main/scala-2.11/org/apache/james/gatling/jmap/JmapMailboxes.scala]: you should not use the default parameter for that but use a factory method instead

File: src/main/scala-2.11/org/apache/james/gatling/jmap/JmapMailboxes.scala [hunk 2]
```diff
@@ -21,6 +21,14 @@ object JmapMailboxes {
     JmapAuthentication.authenticatedQuery("getMailboxes", "/jmap")
       .body(StringBody("""[["getMailboxes", {}, "#0"]]"""))
 
+  def getMailboxIdByName(name: Name) =
+    JmapAuthentication.authenticatedQuery("getMailboxes", "/jmap")
+      .body(StringBody("""[["getMailboxes", {}, "#0"]]"""))
+      .check(getMailboxIdCheck(name))
+
+  def getMailboxIdCheck(name: Name) =
```
Comment (mbaechler) [inline src/main/scala-2.11/org/apache/james/gatling/jmap/JmapMailboxes.scala]: it's not a getter, it extracts and saves

File: src/main/scala-2.11/org/apache/james/gatling/jmap/JmapMailboxes.scala [hunk 3]
```diff
@@ -21,6 +21,14 @@ object JmapMailboxes {
     JmapAuthentication.authenticatedQuery("getMailboxes", "/jmap")
       .body(StringBody("""[["getMailboxes", {}, "#0"]]"""))
 
+  def getMailboxIdByName(name: Name) =
+    JmapAuthentication.authenticatedQuery("getMailboxes", "/jmap")
```
Comment (mbaechler) [inline src/main/scala-2.11/org/apache/james/gatling/jmap/JmapMailboxes.scala]: it's a copy/paste of getMailboxes

File: src/main/scala-2.11/org/apache/james/gatling/jmap/JmapMessages.scala [hunk 4]
```diff
@@ -42,6 +44,24 @@ object JmapMessages {
           "#0"
           ]]"""))
   
+  def getMessageIds(mailboxId: String) =
```
Comment (mbaechler) [inline src/main/scala-2.11/org/apache/james/gatling/jmap/JmapMessages.scala]: still not a getter

File: src/main/scala-2.11/org/apache/james/gatling/jmap/JmapMessages.scala [hunk 5]
```diff
@@ -62,6 +62,23 @@ object JmapMessages {
     jsonPath(messageIdsPath).findAll.saveAs("messageIds")
   }
 
+  def moveSentMessagesToMailboxId = 
+    foreach("${messageIds}", "messageId") {
```
Comment (mbaechler) [inline src/main/scala-2.11/org/apache/james/gatling/jmap/JmapMessages.scala]: how do you know messageIds attribute is sent messages ?

Comment (mbaechler) [inline src/main/scala-2.11/org/apache/james/gatling/jmap/JmapMessages.scala]: why do you want to issue a query for each message ?

File: src/test/scala-2.11/org/apache/james/gatling/jmap/scenari/common/CommonSteps.scala [hunk 6]
```diff
@@ -36,6 +39,23 @@ object CommonSteps {
       }
       .pause(30 second)
 
+  def provisionUsersWithMailboxesAndMessages(users: Seq[Future[User]]) =
+    scenario("ProvisionUsersWithMailboxesAndMessages")
+      .exec(provisionSystemMailboxes(users))
+      .repeat(NumberOfMailboxes) {
```
Comment (mbaechler) [inline src/test/scala-2.11/org/apache/james/gatling/jmap/scenari/common/CommonSteps.scala]: please pass it in parameter

File: src/test/scala-2.11/org/apache/james/gatling/jmap/scenari/JmapBigSetScenario.scala [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+/*
+ * The aim of the scenario is to provide multiple mailboxes per user and several mails in them.
+ * Then, it's just checking that the created mailboxes exists.
+ */
+class JmapBigSetScenario extends Simulation {
+
+  val users = new UserCreator(BaseJamesWebAdministrationUrl).createUsersWithInboxAndOutbox(UserCount)
+  
+  val scn = scenario("JmapBigSet")
```
Comment (mbaechler) [inline src/test/scala-2.11/org/apache/james/gatling/jmap/scenari/JmapBigSetScenario.scala:28]: you don't check mailboxes ?

File: src/main/scala-2.11/org/apache/james/gatling/jmap/JmapMailboxes.scala (hunk 2 above)
Comment (strokyl) [reply src/main/scala-2.11/org/apache/james/gatling/jmap/JmapMailboxes.scala]: I don't understand this function on what json does jsonPath work on ?

**Review by quynhn — COMMENTED** [review]
:+1:

File: src/test/scala-2.11/org/apache/james/gatling/jmap/scenari/common/CommonSteps.scala [hunk 8]
```diff
@@ -36,6 +39,23 @@ object CommonSteps {
...
+        var mailboxName = Name()
+        exec(JmapMailboxes.createMailbox(Id(), mailboxName))
+          .pause(1 second, 2 seconds)
```
Comment (strokyl) [inline src/test/scala-2.11/org/apache/james/gatling/jmap/scenari/common/CommonSteps.scala]: What does that mean you pause for 1 second, 2 seconds or 3 seconds. Or something random between 1 and 2 seconds ?

Comment (chibenwa) [reply src/test/scala-2.11/org/apache/james/gatling/jmap/scenari/common/CommonSteps.scala]: Random between 1 and 2

File: src/main/scala-2.11/org/apache/james/gatling/jmap/JmapMailboxes.scala (hunk 1 above)
Comment (aduprat) [reply src/main/scala-2.11/org/apache/james/gatling/jmap/JmapMailboxes.scala]: There's the same pattern in JmapMessages, I think the whole project needs some refactoring.

File: src/main/scala-2.11/org/apache/james/gatling/jmap/JmapMessages.scala (hunk 5 above)
Comment (aduprat) [reply src/main/scala-2.11/org/apache/james/gatling/jmap/JmapMessages.scala]: That's not really easy, _foreach_ is a gatling method & I don't succeed to chain to something else than gatling methods.

File: src/main/scala-2.11/org/apache/james/gatling/jmap/JmapMailboxes.scala [hunk 9]
```diff
@@ -5,6 +5,19 @@ import io.gatling.http.Predef._
 import io.gatling.http.check.HttpCheck
 import org.apache.james.gatling.utils.JmapChecks
 import org.apache.james.gatling.utils.RetryAuthentication.execWithRetryAuthentication
+import org.apache.james.gatling.utils.RandomStringGenerator
+
+object IdFactory {
```
Comment (mbaechler) [inline src/main/scala-2.11/org/apache/james/gatling/jmap/JmapMailboxes.scala]: I would call it `Id` and call the method `generate`

File: src/main/scala-2.11/org/apache/james/gatling/jmap/JmapMessages.scala [hunk 10]
```diff
@@ -42,6 +44,41 @@ object JmapMessages {
           "#0"
           ]]"""))
   
+  def retrieveMessageIds(mailboxId: String) =
```
Comment (mbaechler) [inline src/main/scala-2.11/org/apache/james/gatling/jmap/JmapMessages.scala]: don't we have a Mailbox.Id type ?

**Review by strokyl — APPROVED** [review]

File: src/main/scala-2.11/org/apache/james/gatling/jmap/JmapMessages.scala [hunk 11]
```diff
@@ -42,6 +44,41 @@ object JmapMessages {
...
+          "#0"
+          ]]"""))
+      .check(saveMessageIds)    
+
+  def saveMessageIds = {
+    jsonPath(messageIdsPath).findAll.saveAs("messageIds")
+  }
+
+  def moveMessagesToMailboxId =
+    foreach("${messageIds}", "messageId") {
```
Comment (mbaechler) [inline src/main/scala-2.11/org/apache/james/gatling/jmap/JmapMessages.scala]: something like that (did not compile, it's just pseudocode) : 
```
exec((s: Session) => s.setAttribute("update", 
  s("messageIds").map(x => """
  "$${x}" : {
                  "mailboxIds": [ "$${mailboxId}" ]"""
).join(","))
.exec(JmapAuthentication.authenticatedQuery("moveSentMessagesToMessageId", "/jmap")
        .body(StringBody(
          s"""[[
            "setMessages",
            {
              "update": {
                "$${update}
                }
              }
            },
            "#0"
            ]]""")))
```

Comment (aduprat) [reply src/main/scala-2.11/org/apache/james/gatling/jmap/JmapMessages.scala]: Thx

File: src/main/scala-2.11/org/apache/james/gatling/jmap/JmapMessages.scala [hunk 12]
```diff
@@ -63,21 +64,22 @@ object JmapMessages {
...
-                "$${messageId}" : {
-                  "mailboxIds": [ "$${mailboxId}" ]
-                }
-              }
-            },
-            "#0"
-            ]]""")))
-    }
+    exec((session: Session) => session.set("update", {
+        var mailboxId = session("mailboxId").as[String]
```
Comment (mbaechler) [inline src/main/scala-2.11/org/apache/james/gatling/jmap/JmapMessages.scala]: val !!!!

**Comment by aduprat** [comment]
#37

**Pull request closed without merging**